src/isolation/landlock.rs

```diff
@@ -0,0 +1,384 @@
+fn is_file(rawfd: RawFd) -> bool {
+	unsafe {
+		let mut stat = std::mem::zeroed();
+		match libc::fstat(rawfd, &mut stat) {
+			0 => (stat.st_mode & libc::S_IFMT) != libc::S_IFDIR,
+			// Should not be practically reachable under optimal circumstances,
+			// but we'll assume otherwise.
+			_ => panic!("stat against fd {:?} failed", rawfd),
+		}
+	}
+}
+
+/// Initializes Landlock using parameters provided by [`UhyveVm::new`](crate::vm::UhyveVm::new).
+///
+/// If the file does not exist, we add the parent directory instead. This might have practical
+/// security implications, however, combined with the other security measures implemented into
+/// Uhyve, this should be fine.
+///
+/// * `sandbox_mode` - File isolation sandbox mode set by the user
+/// * `kernel_path` - Location of the unikernel image
+/// * `output` - Output of Uhyve, used for adding file outputs to Landlock (if applicable)
+/// * `host_paths` - List of host paths derived from mappings by the user
+/// * `temp_dir` - Location of the temporary directory for unmapped files
+pub(crate) fn initialize(
+	sandbox_mode: FileSandboxMode,
+	kernel_path: String,
+	output: &crate::params::Output,
+	host_paths: &mut [OsString],
+	temp_dir: String,
+) -> UhyveLandlockWrapper {
+	// This segment adds certain paths necessary for Uhyve to function before we
+	// enforce Landlock, such as the kernel path and a couple of paths useful for sysinfo.
+	//
+	// See: https://github.com/GuillaumeGomez/sysinfo/blob/8fd58b8/src/unix/linux/cpu.rs#L420
+	let uhyve_ro_paths = [
+		kernel_path,
+		String::from("/etc/"),
+		String::from("/sys/devices/system"),
+		String::from("/proc/cpuinfo"),
+		String::from("/proc/stat"),
+	]
+	.to_vec();
+
+	let mut uhyve_rw_paths: Vec<String> = [String::from("/dev/kvm")].to_vec();
+	let mut uhyve_ro_dirs = Vec::<String>::new();
+
+	host_paths.iter_mut().for_each(|host_path| {
+		let host_pathbuf = PathBuf::from(host_path.clone());
+		if host_pathbuf.exists() {
+			uhyve_rw_paths.push(host_pathbuf.display().to_string());
+			if let Some(parent_dir) = PathBuf::from(host_pathbuf).parent() {
+				uhyve_ro_dirs.push(parent_dir.display().to_string());
+			}
+		} else {
+			uhyve_rw_paths.push(
+				get_parent_directory(host_path)
+					.unwrap()
+					.display()
+					.to_string(),
+			);
+		}
+	});
+
+	if let crate::params::Output::File(path) = output {
+		uhyve_rw_paths.push(path.display().to_string());
+	}
+
+	uhyve_rw_paths.push(temp_dir.clone());
+	if let Some(tmp) = PathBuf::from(temp_dir).parent() {
+		uhyve_ro_dirs.push(tmp.display().to_string());
+	}
+
+	UhyveLandlockWrapper::new(
+		sandbox_mode,
+		&uhyve_rw_paths,
+		&uhyve_ro_paths,
+		&uhyve_ro_dirs,
+	)
+}
+
+#[cfg(test)]
+mod tests {
+	use std::panic;
+
+	use super::*;
+
+	#[test]
+	fn test_get_parent_directory() {
+		assert_eq!(
+			get_parent_directory(&OsString::from("/dev/zero"))
+				.unwrap()
+				.display()
+				.to_string(),
+			"/dev/zero".to_string()
+		);
```

Comment:
I think this behavior is a little counterintuitive, as `/dev/zero` isn't a directory

Reply:
`get_file_or_parent`?

Reply:
The KISS principle strikes again! I'll adjust it real quick once I fix a few technical problems on my end, it hasn't been my lucky week with computers so far. :)